provider/errors.go

```diff
@@ -17,3 +18,34 @@ type ErrUnableToConvertFeatureID struct {
 func (e ErrUnableToConvertFeatureID) Error() string {
 	return fmt.Sprintf("unable to convert feature id %+v to uint64", e.val)
 }
+
+// ErrProviderAlreadyExists is returned when the Provider being registered
+// already exists in the registration system
+type ErrProviderAlreadyExists struct {
+	Name string
+}
+
+func (err ErrProviderAlreadyExists) Error() string {
+	return fmt.Sprintf("provider %s already exists", err.Name)
+}
+
+// ErrUnknownProvider is retured when no providers are registered or requested
+// provider is not registered
+type ErrUnknownProvider struct {
+	Name               string
+	KnownProvidersFunc func() []string
```

Comment:
Seems strange to me to embed a function like this into an error. Maybe consider changing this to `KnownProviders []string` and setting the value at the time of the error.

Reply:
The reason for this, is to lazy load the known providers, instead of have a list passed around the whole time.

Reply:
But why not just call Drivers() right when the error is created and put the list of known drivers at that time in the error? I know it's basically an impossible case right now, but technically the list of drivers could change between the time the error is generated and the Error() method is called.

Reply:
That's a fair point